crates/trie/trie/src/updates.rs

```diff
@@ -2,276 +2,287 @@ use crate::{
-    Delete,
-    /// Update the node entry with the provided value.
-    Update(BranchNodeCompact),
-}
+    /// Finalize state trie updates.
+    pub fn finalize<C>(
+        &mut self,
+        walker: TrieWalker<C>,
+        hash_builder: HashBuilder,
+        destroyed_accounts: HashSet<B256>,
+    ) {
+        // Retrieve deleted keys from trie walker.
+        let (_, removed_node_keys) = walker.split();
+        self.removed_nodes.extend(removed_node_keys);
 
-impl TrieOp {
-    /// Returns `true` if the operation is an update.
-    pub const fn is_update(&self) -> bool {
-        matches!(self, Self::Update(..))
-    }
+        // Retrieve updated nodes from hash builder.
+        let (_, updated_nodes) = hash_builder.split();
+        self.account_nodes.extend(updated_nodes);
 
-    /// Returns reference to updated branch node if operation is [`Self::Update`].
-    pub const fn as_update(&self) -> Option<&BranchNodeCompact> {
-        if let Self::Update(node) = &self {
-            Some(node)
-        } else {
-            None
+        // Add deleted storage tries for destroyed accounts.
+        for destroyed in destroyed_accounts {
+            self.storage_tries.entry(destroyed).or_default().set_deleted(true);
         }
     }
 
-    /// Returns owned updated branch node if operation is [`Self::Update`].
-    pub fn into_update(self) -> Option<BranchNodeCompact> {
-        if let Self::Update(node) = self {
-            Some(node)
-        } else {
-            None
+    /// Converts trie updates into [`TrieUpdatesSorted`].
+    pub fn into_sorted(self) -> TrieUpdatesSorted {
+        let mut account_nodes = Vec::from_iter(self.account_nodes);
+        account_nodes.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+        let mut storage_tries = Vec::from_iter(
+            self.storage_tries
+                .into_iter()
+                .map(|(hashed_address, updates)| (hashed_address, updates.into_sorted())),
+        );
+        storage_tries.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+        TrieUpdatesSorted { removed_nodes: self.removed_nodes, account_nodes, storage_tries }
+    }
+
+    /// Flush updates all aggregated updates to the database.
+    ///
+    /// # Returns
+    ///
+    /// The number of storage trie entries updated in the database.
+    pub fn write_to_database(
+        self,
+        tx: &(impl DbTx + DbTxMut),
+    ) -> Result<usize, reth_db::DatabaseError> {
+        if self.is_empty() {
+            return Ok(0)
         }
+
+        // Track the number of inserted entries.
+        let mut num_entries = 0;
+
+        // Merge updated and removed nodes. Updated nodes must take precedence.
+        let capacity = self.removed_nodes.len() + self.account_nodes.len();
+        let mut account_updates = Vec::with_capacity(capacity);
+        for nibbles in self.removed_nodes {
+            if !self.account_nodes.contains_key(&nibbles) {
+                account_updates.push((nibbles, None));
+            }
```

Comment:
this means we may end up allocating more than necessary? by how much?

maybe filter + collect + extend?